File: packages/firebase_crashlytics/darwin/Classes/FirebaseCrashlyticsPlugin.m
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FirebaseCrashlyticsPlugin.h"

#import <Firebase/Firebase.h>

@interface FirebaseCrashlyticsPlugin ()
@property(nonatomic, retain) FlutterMethodChannel *channel;
@end

@implementation FirebaseCrashlyticsPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel =
      [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/firebase_crashlytics"
                                  binaryMessenger:[registrar messenger]];
  FirebaseCrashlyticsPlugin *instance = [[FirebaseCrashlyticsPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];

  [FIRApp configure];


  SEL sel = NSSelectorFromString(@"registerLibrary:withVersion:");
  if ([FIRApp respondsToSelector:sel]) {
    [FIRApp performSelector:sel withObject:LIBRARY_NAME withObject:LIBRARY_VERSION];
  }
}

- (instancetype)init {
  self = [super init];
  if (self) {
    if (![FIRApp defaultApp]) {
      [FIRApp configure];
    }
  }
  return self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  if ([@"Crashlytics#onError" isEqualToString:call.method]) {
    // Add logs.
    NSArray *logs = call.arguments[@"logs"];
    for (NSString *log in logs) {
      // Here and below, use CLSLog instead of CLS_LOG to try and avoid
      // automatic inclusion of the current code location. It also ensures that
      // the log is only written to Crashlytics and not also to the offline log
      // as explained here:
      // https://support.crashlytics.com/knowledgebase/articles/92519-how-do-i-use-logging
        [[FIRCrashlytics crashlytics] logWithFormat:@"%@", log];
    }

    // Set keys.
    NSArray *keys = call.arguments[@"keys"];
    for (NSDictionary *key in keys) {
      if ([@"int" isEqualToString:key[@"type"]]) {
          [[FIRCrashlytics crashlytics] setCustomValue:@((int)call.arguments[@"value"]) forKey:call.arguments[@"key"]];
      } else if ([@"double" isEqualToString:key[@"type"]]) {
          [[FIRCrashlytics crashlytics] setCustomValue:@([call.arguments[@"value"] floatValue]) forKey:call.arguments[@"key"]];
      } else if ([@"string" isEqualToString:key[@"type"]]) {
           [[FIRCrashlytics crashlytics] setCustomValue:call.arguments[@"value"] forKey:call.arguments[@"key"]];
      } else if ([@"boolean" isEqualToString:key[@"type"]]) {
          [[FIRCrashlytics crashlytics] setCustomValue:@([call.arguments[@"value"] boolValue]) forKey:call.arguments[@"key"]];
      }
    }

    // Add additional information from the Flutter framework to the exception reported in
    // Crashlytics.
    NSString *information = call.arguments[@"information"];
    if ([information length] != 0) {
      [[FIRCrashlytics crashlytics] logWithFormat:@"%@", information];
    }

    // Report crash.
    NSArray *errorElements = call.arguments[@"stackTraceElements"];
    NSMutableArray *frames = [NSMutableArray array];
    for (NSDictionary *errorElement in errorElements) {
      [frames addObject:[self generateFrame:errorElement]];
    }

    NSString *context = call.arguments[@"context"];
    NSString *reason;
    if (context != nil) {
      reason = [NSString stringWithFormat:@"thrown %@", context];
    }

    FIRExceptionModel *model = [[FIRExceptionModel alloc] initWithName:call.arguments[@"exception"] reason:reason];
      model.stackTrace = errorElements;

    result(@"Error reported to Crashlytics.");
  } else if ([@"Crashlytics#isDebuggable" isEqualToString:call.method]) {
    result([NSNumber numberWithBool:NO]);
  } else if ([@"Crashlytics#getVersion" isEqualToString:call.method]) {
    result(@"6.23.0");
  } else if ([@"Crashlytics#setUserEmail" isEqualToString:call.method]) {
    [[FIRCrashlytics crashlytics] setCustomValue:call.arguments[@"email"] forKey:@"email"];
    result(nil);
  } else if ([@"Crashlytics#setUserName" isEqualToString:call.method]) {
    [[FIRCrashlytics crashlytics] setCustomValue:call.arguments[@"name"] forKey:@"name"];
    result(nil);
  } else if ([@"Crashlytics#setUserIdentifier" isEqualToString:call.method]) {
    [[FIRCrashlytics crashlytics] setUserID:call.arguments[@"identifier"]];
    result(nil);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (FIRStackFrame *)generateFrame:(NSDictionary *)errorElement {
  return [FIRStackFrame stackFrameWithSymbol:[errorElement valueForKey:@"method"] file:[errorElement valueForKey:@"file"] line:[[errorElement valueForKey:@"line"] intValue]];
}

@end

File: packages/firebase_crashlytics/android/src/main/java/io/flutter/plugins/firebase/crashlytics/firebasecrashlytics/FirebaseCrashlyticsPlugin.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebase.crashlytics.firebasecrashlytics;

import android.content.Context;
import android.util.Log;

import com.google.firebase.crashlytics.FirebaseCrashlytics;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** FirebaseCrashlyticsPlugin */
public class FirebaseCrashlyticsPlugin implements FlutterPlugin, MethodCallHandler {
  public static final String TAG = "CrashlyticsPlugin";
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(FlutterPluginBinding binding) {
    BinaryMessenger binaryMessenger = binding.getBinaryMessenger();
    channel = setup(binaryMessenger, binding.getApplicationContext());
  }

  @Override
  public void onDetachedFromEngine(FlutterPluginBinding binding) {
    if (channel != null) {
      channel.setMethodCallHandler(null);
      channel = null;
    }
  }

  private static MethodChannel setup(BinaryMessenger binaryMessenger, Context context) {
    final MethodChannel channel =
            new MethodChannel(binaryMessenger, "plugins.flutter.io/firebase_crashlytics");
    channel.setMethodCallHandler(new FirebaseCrashlyticsPlugin());
    
    return channel;
  }

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    setup(registrar.messenger(), registrar.context());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("Crashlytics#onError")) {
      // Add logs.
      List<String> logs = call.argument("logs");
      for (String log : logs) {
        FirebaseCrashlytics.getInstance().log(log);
      }

      // Set keys.
      List<Map<String, Object>> keys = call.argument("keys");
      for (Map<String, Object> key : keys) {
        switch ((String) key.get("type")) {
          case "int":
            FirebaseCrashlytics.getInstance().setCustomKey((String) key.get("key"), (int) key.get("value"));
            break;
          case "double":
            FirebaseCrashlytics.getInstance().setCustomKey((String) key.get("key"), (double) key.get("value"));
            break;
          case "string":
            FirebaseCrashlytics.getInstance().setCustomKey((String) key.get("key"), (String) key.get("value"));
            break;
          case "boolean":
            FirebaseCrashlytics.getInstance().setCustomKey((String) key.get("key"), (boolean) key.get("value"));
            break;
        }
      }

      // Report crash.
      String dartExceptionMessage = (String) call.argument("exception");
      Exception exception = new Exception(dartExceptionMessage);
      List<Map<String, String>> errorElements = call.argument("stackTraceElements");
      List<StackTraceElement> elements = new ArrayList<>();
      for (Map<String, String> errorElement : errorElements) {
        StackTraceElement stackTraceElement = generateStackTraceElement(errorElement);
        if (stackTraceElement != null) {
          elements.add(stackTraceElement);
        }
      }
      exception.setStackTrace(elements.toArray(new StackTraceElement[elements.size()]));

      FirebaseCrashlytics.getInstance().setCustomKey("exception", (String) call.argument("exception"));

      // Set a "reason" (to match iOS) to show where the exception was thrown.
      final String context = call.argument("context");
      if (context != null) FirebaseCrashlytics.getInstance().setCustomKey("reason", "thrown " + context);

      // Log information.
      final String information = call.argument("information");
      if (information != null && !information.isEmpty()) FirebaseCrashlytics.getInstance().log(information);

      FirebaseCrashlytics.getInstance().recordException(exception);
      result.success("Error reported to Crashlytics.");
    } else if (call.method.equals("Crashlytics#isDebuggable")) {
      result.success(false);
    } else if (call.method.equals("Crashlytics#getVersion")) {
      result.success("firebase");
    } else if (call.method.equals("Crashlytics#setUserEmail")) {
      FirebaseCrashlytics.getInstance().setCustomKey("email",(String) call.argument("email"));
      result.success(null);
    } else if (call.method.equals("Crashlytics#setUserIdentifier")) {
      FirebaseCrashlytics.getInstance().setUserId((String) call.argument("identifier"));
      result.success(null);
    } else if (call.method.equals("Crashlytics#setUserName")) {
      FirebaseCrashlytics.getInstance().setCustomKey("name",(String) call.argument("name"));
      result.success(null);
    } else {
      result.notImplemented();
    }
  }

  /**
   * Extract StackTraceElement from Dart stack trace element.
   *
   * @param errorElement Map representing the parts of a Dart error.
   * @return Stack trace element to be used as part of an Exception stack trace.
   */
  private StackTraceElement generateStackTraceElement(Map<String, String> errorElement) {
    try {
      String fileName = errorElement.get("file");
      String lineNumber = errorElement.get("line");
      String className = errorElement.get("class");
      String methodName = errorElement.get("method");

      return new StackTraceElement(
              className == null ? "" : className, methodName, fileName, Integer.parseInt(lineNumber));
    } catch (Exception e) {
      Log.e(TAG, "Unable to generate stack trace element from Dart side error.");
      return null;
    }
  }
}
